### PR TITLE
Make pkg/resource expose an interface

### PR DESCRIPTION
**Pull request by qbarrand**
This PR:
- adds the `Creator` interface to `pkg/resource`, as well as a default implementation. Although this package had many public functions, none besides `CreateFromYAML` seem to be used outside. I decided to leave them public for now to preserve the unit tests that target them, otherwise the coverage would be greatly reduced.
- adds unit tests to the only `Creator` method. Coverage is still very low because that method has many branches and writing many tests cases is not practical. We should eventually refactor this package to add more mockable types.

Adding `Creator` as an interface allows us to inject it as a dependency in `pkg/helmer`. Work ongoing there in #189.

/cc @pacevedom @yevgeny-shnaidman @ybettan @pmtk
/ok-to-test

### CODE REVIEW

File: pkg/resource/resource.go [hunk 1]
```diff
@@ -33,136 +33,74 @@ type resourceCallbacks map[string]func(obj *unstructured.Unstructured, sr interf
...
+		kubeClient:    kubeClient,
 	}
-	return false
-
 }
 
-func UpdateResourceVersion(req *unstructured.Unstructured, found *unstructured.Unstructured) error {
-
-	kind := found.GetKind()
+func (c *creator) AfterCRUD(obj *unstructured.Unstructured, namespace string) error {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:67]: This function is used only in resource package, lets make it private

File: pkg/resource/resource.go [hunk 2]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
+func IsNotUpdateable(kind string) bool {
+	// ServiceAccounts cannot be updated, maybe delete and create?
+	if kind == "ServiceAccount" || kind == "Pod" {
+		return true
+	}
+	return false
+}
+
+// Some resources need an updated resourceversion, during updates
+func NeedsResourceVersionUpdate(kind string) bool {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:498]: lets also make it a private function

File: pkg/resource/resource.go [hunk 3]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
+		kind == "ValidatingWebhookConfiguration" ||
+		kind == "Deployment" ||
+		kind == "ImagePolicy" {
+		return true
+	}
+	return false
+
+}
+
+func UpdateResourceVersion(req *unstructured.Unstructured, found *unstructured.Unstructured) error {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:525]: lets make it a private function

File: pkg/resource/resource.go [hunk 4]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
+
+		if err = unstructured.SetNestedField(req.Object, clusterIP, "spec", "clusterIP"); err != nil {
+			return fmt.Errorf("couldn't update clusterIP: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func SetNodeSelectorTerms(obj *unstructured.Unstructured, terms map[string]string) error {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:554]: lets make it a private function

File: pkg/resource/resource.go [hunk 5]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
+	}
+
+	if err = unstructured.SetNestedMap(obj.Object, nodeSelector, fields...); err != nil {
+		return fmt.Errorf("cannot update nodeSelector for %s : %w", obj.GetName(), err)
+	}
+
+	return nil
+}
+
+func IsOneTimer(obj *unstructured.Unstructured) (bool, error) {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:599]: lets make it a private function

File: pkg/resource/resource.go [hunk 6]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
+
+		if restartPolicy == "Never" {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func SetMetaData(obj *unstructured.Unstructured, nm string, ns string) {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:616]: lets make a private function

File: pkg/resource/resource.go [hunk 7]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	labels["app.kubernetes.io/managed-by"] = "Helm"
+
+	obj.SetLabels(labels)
+}
+
+func BeforeCRUD(obj *unstructured.Unstructured, sr interface{}) error {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:640]: lets make a private function

File: pkg/resource/resource.go [hunk 8]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
+		kind == "ClusterRole" ||
+		kind == "ClusterRoleBinding" ||
+		kind == "SecurityContextConstraint" ||
+		kind == "SpecialResource" {
+		return false
+	}
+	return true
+}
+
+func IsNotUpdateable(kind string) bool {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:489]: let make it a private function

File: pkg/resource/resource.go [hunk 9]
```diff
@@ -632,8 +462,194 @@ func sendNodesMetrics(obj *unstructured.Unstructured, crName string) {
...
 		nodesStr := strings.Join(nodesNames, ",")
-		metricsClient.SetUsedNodes(crName, obj.GetName(), obj.GetNamespace(), kind, nodesStr)
+		c.metricsClient.SetUsedNodes(crName, obj.GetName(), obj.GetNamespace(), kind, nodesStr)
 	} else {
-		log.Info("No assigned nodes for found for UsedNodes metric", "kind", kind, "name", obj.GetName(), "crName", crName)
+		c.log.Info("No assigned nodes for found for UsedNodes metric", "kind", kind, "name", obj.GetName(), "crName", crName)
+	}
+}
+
+func IsNamespaced(kind string) bool {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:478]: lets make it a private function

File: pkg/resource/resource.go [hunk 10]
```diff
@@ -197,119 +135,14 @@ func CreateFromYAML(yamlFile []byte,
...
-			return true, nil
-		}
-	}
-
-	return false, nil
-}
-
 // CRUD Create Update Delete Resource
-func CRUD(obj *unstructured.Unstructured, releaseInstalled bool, owner v1.Object, name string, namespace string) error {
+func (c *creator) CRUD(obj *unstructured.Unstructured, releaseInstalled bool, owner v1.Object, name string, namespace string) error {
```
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:146]: lets make it a private function

File: pkg/resource/resource.go (hunk 7 above)
Comment (yevgeny-shnaidman) [inline pkg/resource/resource.go:640]: Also, some of the private functions are members of the creator struct, and some are just package function. WDYT about making them all members of the creator?  Right now we have AfterCrud as a member, and Before Crud as not, which is kind of confusisng

File: main.go [hunk 11]
```diff
@@ -88,16 +89,14 @@ func main() {
 
 	metricsClient := metrics.New()
 
-	resource.RuntimeScheme = mgr.GetScheme()
-
-	creator := resource.NewCreator(metricsClient, clients.Interface)
+	creator := resource.NewCreator(clients.Interface, metricsClient, poll.New(), scheme)
```
Comment (yevgeny-shnaidman) [inline main.go:92]: Won't it make a code a little bit confusing? Maybe we can change all the packages interfaces initializations in one final PRs, once all the packages has been converted into typed interface?

File: pkg/resource/resource.go (hunk 1 above)
Comment (qbarrand) [reply pkg/resource/resource.go:67]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 2 above)
Comment (qbarrand) [reply pkg/resource/resource.go:498]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 3 above)
Comment (qbarrand) [reply pkg/resource/resource.go:525]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 4 above)
Comment (qbarrand) [reply pkg/resource/resource.go:554]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 5 above)
Comment (qbarrand) [reply pkg/resource/resource.go:599]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 7 above)
Comment (qbarrand) [reply pkg/resource/resource.go:640]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 8 above)
Comment (qbarrand) [reply pkg/resource/resource.go:489]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 9 above)
Comment (qbarrand) [reply pkg/resource/resource.go:478]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 10 above)
Comment (qbarrand) [reply pkg/resource/resource.go:146]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: pkg/resource/resource.go (hunk 7 above)
Comment (qbarrand) [reply pkg/resource/resource.go:640]: We agreed to address this with https://issues.redhat.com/browse/MGMT-8567

File: main.go (hunk 11 above)
Comment (qbarrand) [reply main.go:92]: I tried adding a global variable to `pkg/resource`, but not initializing it from `main` proved difficult. Let's leave it like this for now, `pkg/helmer` refactoring is next in line anyway.

**Pull request merged**